[#25] 리뷰 엔티티 관련 기능 추가

canlion opened the pull request:
# 목표
* 리뷰 엔티티 관련 기능 추가
* sync, async, blocking, non-blocking
  * 스프링 로직의 동기, 비동기에 따른 차이는 무엇인지?
  * 내 프로젝트에 어떻게 적용할 수 있을지?

canlion committed:
fix: @Repository 제거

- 인터페이스에 설정된 어노테이션 제거

canlion committed:
feat: 음식 조회 기능 추가

- 음식 레포지터리에 누락된 음식 조회 기능 추가

canlion committed:
feat: 리뷰 엔티티 추가

canlion committed:
feat: 리뷰 레포지터리와 엔티티 조회 예외 추가

- 리뷰 레포지터리 인터페이스와 구현체 추가
  - 리뷰 추가/조회/삭제
  - 가게/작성회원/음식으로 리뷰 조회
- 엔티티 조회 예외 ReviewNotExistException 추가

canlion committed:
feat: 음식 엔티티 조회 예외 추가

- 엔티티 조회 예외 FoodNotExistException 추가

canlion committed:
feat: 엔티티 조회를 수행하는 서비스 추가

- 순환 참조와 코드 중복을 피하기 위해 서비스를 참조되는 역할, 참조하는
  역할로 분할
  - 다른 엔티티의 서비스를 통해 엔티티 조회시 순환 참조 위험 존재
  - 레포지터리를 직접 참조하면 엔티티 조회 및 예외 발생 코드가 중복
  - 엔티티 조회 기능을 별개의 서비스로 추출해 다른 서비스에서 참조

canlion committed:
fix: 누락된 엔티티 id 생성 전략 적용

canlion committed:
feat: 리뷰 서비스 추가

- 리뷰 추가, 제거, 조회 기능 제공

canlion committed:
feat: 리뷰 컨트롤러 추가

- 리뷰 등록, 조회 기능 제공

canlion commented:
## 리뷰 엔티티 관련 기능 추가

그간 작성했던 코드들을 기반으로 리뷰 엔티티 관련 기능을 추가했습니다.

```java
@Entity
public class Review {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long reviewId;
    @ManyToOne
    private Member member;
    @ManyToOne
    private Store store;
    @ManyToOne
    private Food food;
    private String content;
}
```

### 레포지터리

SpringDataJPA 이용해 간단히 작성했습니다.

### 서비스

```java
@Service
@RequiredArgsConstructor
public class ReviewLookupService {

    private final ReviewRepository repository;

    public Review findByReviewId(Long reviewId) {
        return repository.findByReviewId(reviewId).orElseThrow(() -> new ReviewNotExistException(reviewId));
    }
}

```
```java
@Slf4j
@Service
@RequiredArgsConstructor
public class ReviewService {

    private final ReviewRepository repository;

    private final ReviewLookupService reviewLookupService;
    private final MemberLookupService memberLookupService;
    private final StoreLookupService storeLookupService;
    private final FoodLookupService foodLookupService;

    public void addReview(Review review, Long memberId, Long foodId) {
        // TODO: 회원이 CUSTOMER 타입인지 확인, 아닌 경우 예외
        Food food = foodLookupService.findByFoodId(foodId);
        review.setFood(food);
        review.setStore(food.getStore());
        review.setMember(memberLookupService.findByMemberId(memberId));
        repository.save(review);
        log.debug("review added: {}", review);
    }
```

#### 엔티티 조회만을 담당하는 서비스 클래스 작성

 리뷰 엔티티는 작성한 회원, 리뷰 대상 음식과 판매 가게와 연관되어 여러 엔티티 조회가 필요합니다. 서비스에서 레포지터리를 직접 참조하면 엔티티 조회 & 조회 대상이 없다면 예외를 발생시키는 코드가 중복되어 나타나고, 다른 서비스를 참조해 엔티티를 조회하면 기능을 확장하며 순환 참조가 발생할 수 있습니다. 그래서 엔티티를 조회하는 기능만 담당하는 서비스를 만들어 기존 서비스를 다른 서비스를 참조하는 역할과 다른 서비스가 참조하는 역할로 나누어보았습니다.
 클래스가 늘어나지만 단순 조회 역할만 담당하므로 기능이 확장되어도 수정이 일어나지 않을 가능성이 크고, 코드 중복과 순환 참조 위험성을 줄일 수 있는 점이 더 이득이라 생각되어 적용해봤습니다.

### 컨트롤러

 리뷰 추가, 조회 관련 기능만 추가된 상태입니다.

canlion committed:
feat: 리뷰 삭제 기능 추가

* 리뷰 작성자가 리뷰 삭제하는 기능 추가
  - 리뷰 작성자가 아닌 회원이 삭제를 요청하는 경우
    NotReviewAuthorException 예외 발생

canlion committed:
feat: 개별 리뷰 조회 기능 추가

canlion committed:
refactor: 리뷰 삭제 기능의 불필요한 엔티티 조회 제거

* 리뷰 삭제 기능에서 회원 엔티티를 조회하지 않고 바로 삭제 요청 회원
  식별번호와 리뷰 작성 회원의 식별번호를 비교하도록 수정

canlion committed:
refactor: 리뷰 등록 기능의 엔티티 조회를 비동기로 수행하도록 수정

* 리뷰 등록 시 작성 회원, 리뷰 대상 음식 조회를 비동기적으로 수행하도록
  변경
  - 매우 간단한 쿼리이므로 비동기 실행으로 이득이 있을지는 확인이 필요

canlion committed:
[#24] 코드 개선 (#27)

* refactor: 상수 값과 이름 변경

- 상수의 의미를 더 명확히 표현하도록 값과 이름 변경
  - 로그인한 회원 식별 번호 의미: `loginMember` -> `LOGIN_MEMBER_ID`

* refactor: isMemberTypeOwner 메서드를 MemberLookupService로 이동

- MemberService의 isMemberTypeOwner 메서드를 MemberLookupService로 이동.
  - 서비스간의 순환 참조 위험을 줄이기 위해 MemberLookupService 추가.
  - isMemberTypeOwner 메서드는 여러 서비스에서 사용하므로 이동시킴.

* refactor: checkStoreBelongTo 메서드를 StoreLookupService로 이동

- StoreService의 checkStoreBelongTo 메서드를 StoreLookupService로 이동.
  - 서비스간의 순환 참조 위험을 줄이기 위해 StoreLookupService 추가.
  - checkStoreBelongTo 메서드는 여러 서비스에서 사용하므로 이동시킴.

* refactor: 서비스가 lookup 서비스를 통해 엔티티를 조회하도록 변경

- 서비스는 각 엔티티의 lookup 서비스의 엔티티 조회 메서드를 사용
  - 엔티티 조회에 동반되는 예외 발생 코드를 lookup 서비스의 조회
    메서드로 은닉하여 가독성을 향상시킴.

* refactor: 기능이 중복되는 메서드 제거

- StoreService의 checkStoreExistence 메서드 제거
  - 가게 조회 메서드와 동일 기능 메서드이므로 제거하고 가게 조회
    메서드로 대체

* refactor: isMemberTypeOwner 메서드가 조회한 회원 엔티티를 반환

- 회원 타입을 확인하기 위해 조회한 회원 엔티티를 재사용할 수 있도록 반환

* refactor: Optional 활용하여 로그인 로직 가독성 개선

- 중첩 if문을 Optional의 or, map으로 대체함.

* refactor: 로그인 로직의 Optional.map을 filter로 변경

- Optional로 감싼 Member 객체를 변형하지않고 비밀번호 일치 여부만
  확인하므로 map 보다 filter가 적절함.

* refactor: 불필요한 import문 제거

* refactor: FoodAttributeNotExistException을 Food 컨트롤러에서 처리

- Food 컨트롤러의 메서드 호출에서만 발생하는 예외이므로 어드바이스를
  제거하고 Food 컨트롤러에서 처리하도록 변경

* chore: 레포지터리의 HashMap 구현 제거

- MySql 레포지터리로 전환하여 사용하지 않는 코드 제거

* refactor: @Transactional을 메서드 단위로 적용

* 단순 조회 등 트랜잭션이 불필요한 메서드에 트랜잭션을 생성하는 낭비가
  발생하지 않도록 메서드 단위로 @Transactional 적용

* feat: 음식 속성(메뉴 타입, 카테고리, 재료)을 모두 조회하는 기능 추가

* 메뉴 타입, 카테고리, 재료는 음식 등록, 음식 필터링 등의 기능에서 함께
  사용되므로 한 번의 요청으로 모두 반환할 수 있는 기능을 추가.

* refactor: 예외 핸들러 반환값을 DTO로 변경

* 예외 정보를 Map으로 전달하는 대신 DTO로 변경
  - 향후 예외에 대한 추가 정보를 담는 경우처럼 확장성을 고려함.
  - 모든 예외 핸들러의 반환을 DTO로 통일해 일관성 확보하며 Map의 key를
    잘못적는 실수 방지
  - 반환 타입만으로 반환값의 의미를 전달해 가독성 개선

* refactor: DTO -> entity 변환 기능 중 불필요한 if 제거

* 전달된 DTO의 타입에 맞춰 엔티티로 변환하는 메서드에서 불필요한 if를
  제거하고 오버로딩을 이용하도록 수정

* fix: 추가된 기능에 변경된 상수명 반영

* 25 브랜치에서 추가된 기능에 24 브랜치에서의 상수명 변경을 반영

f-lab-lyan commented on api/src/main/java/burgermap/session/SessionConstants.java:
```diff
@@ -1,5 +1,5 @@
 
 public interface SessionConstants {
-    static String loginMember = "loginMember";
+    static String LOGIN_MEMBER_ID = "LOGIN_MEMBER_ID";
```
👍

f-lab-lyan commented on api/src/main/java/burgermap/service/StoreService.java:
```diff
@@ -1,56 +1,51 @@
 @RequiredArgsConstructor
 public class StoreService {
-    private final MemberService memberService;
+    private final MemberLookupService memberLookupService;
```
얘넨 약간 DAO같은 느낌인가보군요!

canlion replied:
~DB에 직접 접근하는 역할은 레포지터리가 담당하고 `LookupService`는 레포지터리를 참조하여 DB관련 작업을 수행하기에 DAO보다는 '다른 서비스 레이어들이 참조하는 서비스'로 봐주시면 될 듯 합니다.~

canlion replied:
수업에서 설명해주신대로 다시 정리를 해보면 DAO는 DB 엑세스와 편의를 위한 로직을 결합한 레이어이며,
`LookupService`는 엔티티 조회 + 조건에 따른 예외 발생 + 일부 엔티티와 관련된 편의 로직(가게 소유 회원 대조, 리뷰 작성 회원 대조, ...)을 포함하므로 DAO라고 생각됩니다.

f-lab-lyan commented on api/src/main/java/burgermap/service/MemberService.java:
이 파일 뿐만 아니라 전반적으로 @Transactional이 늘어난 것 같은데, 어떤 변화가 있었던 걸까요?

canlion replied:
트랜잭션을 클래스 단위로 적용하는 경우 단순 조회 메서드처럼 트랜잭션이 불필요한 기능에서도 트랜잭션을 생성하므로 자원 낭비가 있다고 조언해주셔서 메서드 단위로 적용했습니다.

f-lab-lyan reviewed (approved):
마지막 변경사항만 봤는데, 좋은 것 같네요!!! 👍

f-lab-lyan commented:
그나저나 CI가 실패하는 것 같은데, 이 부분만 해결하고, 머지하고, 다른 PR로 작업해보시죠. :-)

canlion committed:
[#28] 누락된 기본 기능 추가 (#29)

* feat: 음식 조회 기능 추가

* `foods/{foodId}` URL을 통해 음식을 조회하는 기능 추가
* FoodNotExistException이 FoodController와 ReviewController에 걸쳐
  발생하므로 FoodExceptionAdvice를 추가해 예외 처리

* feat: 음식 엔티티 수정 기능 추가

* 요청한 회원이 음식이 등록된 가게 소유주인지 확인 후 음식 엔티티
  수정하는 기능 추가

* fix: 관계 필드를 참조하도록 쿼리 오류 수정

* 쿼리가 Food.store.storeId 대신 Food.storeId를 참조하는 오류를 수정

* feat: 가게 엔티티 존재를 확인하는 기능 추가

* 주어진 ID에 해당하는 가게가 존재하는지 확인하는 기능 추가
  - 객체를 조회하는 findBy~ 대비 효율적: 오브젝트 맵핑 생략

* feat: 특정 가게에 등록된 모든 음식 조회 기능 추가

* feat: 음식 엔티티 삭제 기능 추가

* refactor: 미리 조회한 엔티티를 통해 엔티티를 제거하도록 수정

* 엔티티 제거 권한을 확인하기 위해 미리 조회한 엔티티를 통해 엔티티 제거
  기능이 수행되도록 변경하여 추가적인 엔티티 조회 방지

* feat: 리뷰 수정 기능 추가

* feat: 음식 필터 기능 추가

* 주어진 조건을 통해 음식 필터링: 다음 조건들을 and 조건으로 검색.
  null인 조건은 무시됨.
  - 특정 메뉴 카테고리의 음식 (nullable)
  - 특정 재료들을 모두 포함한 음식 (nullable)

canlion committed:
test: DAO 추가로 인한 테스트 코드 수정

- 서비스레이어 일부 기능이 레포지터리 대신 DAO를 참조하도록 구성 변경
- 일부 테스트 케이스가 레포지터리 대신 DAO를 mocking하도록 수정
